### PR TITLE
[SPARK-27679][SQL] Improve LIKE optimization

### PR DESCRIPTION
**Pull request by Achuth17**
## What changes were proposed in this pull request?

The following improvements are possible when simplifying `LIKE` expressions by adding support for underscore case optimization similar to '%' :

```
    Filter name#10 LIKE _abc → Filter (length(name#10) = 4 && EndsWith(name#10, abc))
    Filter name#10 LIKE abc_ → Filter (length(name#10) = 4 && StartsWith(name#10, abc))
    Filter name#10 LIKE abc_def -> Filter ((length(name#10) = 7) && (StartsWith(name#10, abc) && EndsWith(name#10, def)))
```

## How was this patch tested?

Added UTs.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 1]
```diff
@@ -479,12 +479,15 @@ object LikeSimplification extends Rule[LogicalPlan] {
...
   private val contains = "%([^_%]+)%".r
   private val equalTo = "([^_%]*)".r
+  private val endsWithUnderscore = "([^_%]+)_".r
+  private val startsWithUnderscore = "_([^_%]+)".r
+  private val containsSingleUnderscore = "([^_%]+)_([^_%]+)".r
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
     case Like(input, Literal(pattern, StringType)) =>
       if (pattern == null) {
         // If pattern is null, return null value directly, since "col like null" == null.
```
Comment (venkata91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Need to fix this comment? Why null is not a valid pattern?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 2]
```diff
@@ -479,12 +479,15 @@ object LikeSimplification extends Rule[LogicalPlan] {
...
-        Literal(null, BooleanType)
+        Literal(false, BooleanType)
```
Comment (venkata91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Need to fix this comment? Why null is not a valid pattern?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 3]
```diff
@@ -479,12 +479,15 @@ object LikeSimplification extends Rule[LogicalPlan] {
   private val startsAndEndsWith = "([^_%]+)%([^_%]+)".r
   private val contains = "%([^_%]+)%".r
   private val equalTo = "([^_%]*)".r
+  private val endsWithUnderscore = "([^_%]+)_".r
+  private val startsWithUnderscore = "_([^_%]+)".r
+  private val containsSingleUnderscore = "([^_%]+)_([^_%]+)".r
```
Comment (venkata91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: May be startsAndEndsWithUnderscore or something?

**Comment by Achuth17** [comment]
@dongjoon-hyun @mgaido91 @maropu @gatorsmile, 
Can one of you review this change? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 4]
```diff
@@ -479,12 +479,15 @@ object LikeSimplification extends Rule[LogicalPlan] {
...
+  private val endsWithUnderscore = "([^_%]+)_".r
+  private val startsWithUnderscore = "_([^_%]+)".r
+  private val startsAndEndsWithUnderscore = "([^_%]+)_([^_%]+)".r
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transformAllExpressions {
     case Like(input, Literal(pattern, StringType)) =>
       if (pattern == null) {
         // If pattern is null, return null value directly, since "col like null" == null.
-        Literal(null, BooleanType)
+        Literal(false, BooleanType)
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: why did you change this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 5]
```diff
@@ -479,12 +479,15 @@ object LikeSimplification extends Rule[LogicalPlan] {
   private val startsAndEndsWith = "([^_%]+)%([^_%]+)".r
   private val contains = "%([^_%]+)%".r
   private val equalTo = "([^_%]*)".r
+  private val endsWithUnderscore = "([^_%]+)_".r
+  private val startsWithUnderscore = "_([^_%]+)".r
+  private val startsAndEndsWithUnderscore = "([^_%]+)_([^_%]+)".r
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: this is not a `startsAndEndsWithUnderscore`, but the underscore is inside the pattern

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala [hunk 6]
```diff
@@ -500,6 +503,18 @@ object LikeSimplification extends Rule[LogicalPlan] {
             Contains(input, Literal(infix))
           case equalTo(str) =>
             EqualTo(input, Literal(str))
+          // case: abc_
+          case endsWithUnderscore(prefix) =>
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: what if the pattern is something like `abc\_`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 4 above)
Comment (Achuth17) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: I was planning on changing `like null` to `like false` so that we can change the underlying table read but I realized that `like null` is a valid check. I will revert this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 5 above)
Comment (Achuth17) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 6 above)
Comment (Achuth17) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: I have handled this case and added UTs for this. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala (hunk 2 above)
Comment (Achuth17) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala]: Reverted this change.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Thank you for making a PR. Yes, this has some benefit, but this doesn't support '_' in a general way. Since this is a partial approach, many users will try to use mixed patterns like 'a_b%' and will get confused. I'm worrying the situation which the users file a bug for that mixed `a_b%` during migrating their SQL statements from RDBMS to Spark. We can add some documents for this limitation, but the current improvement of this PR also is achievable by adding length checks with the existing '%'. For now, I'm a little negative for the current partial status of this PR. Sorry about that.

cc @gatorsmile and @cloud-fan

**Comment by cloud-fan** [comment]
Agree with @dongjoon-hyun . This is too special, e.g. `__abc` is not supported.

**Comment by dongjoon-hyun** [comment]
According to the current status, we had better close this PR for now, @Achuth17 .
Please feel free to reopen the PR with a general approach. 
I'll keep SPARK-27679 open for you. Thank you for the contribution(filing the issue)!

**Comment by Achuth17** [comment]
@dongjoon-hyun I could use some input on what a general approach would look like, I will take another swing at the problem.

I can add more general cases like `__abc` or `abc__`, other than that I can't think of anything that would make cases like `ab_c%` work. Can you suggest a good approach?

**Comment by dongjoon-hyun** [comment]
Unfortunately, No. I believe that you already know DBMS behaviors because you are suggesting to have this. Maybe, some open source implementation can help you some ideas if you search them. For me, I didn't take a look at them.
